komposable-architecture/src/main/java/com/toggl/komposable/extensions/EffectExtensions.kt

```diff
@@ -1,62 +1,153 @@
- * @return A list of effects.
+ * @param actions Vararg of actions to be returned by the effect.
+ * @return ReduceResult containing the current state and the specified effect.
  */
-fun <Action> effectOf(vararg effects: Effect<Action>): List<Effect<Action>> =
-    effects.toList()
+fun <State, Action> State.withEffect(vararg actions: Action): ReduceResult<State, Action> =
+    ReduceResult(this, Effect.of(*actions))
 
-infix operator fun <Action> Effect<Action>.plus(otherEffect: Effect<Action>) = listOf(this, otherEffect)
+/**
+ * Returns a ReduceResult with the current state and a flow-based effect.
+ * The effect can be made cancellable using an ID.
+ *
+ * @param flow The flow of actions to be executed as an effect.
+ * @param id Optional ID to make the effect cancellable. If null, the effect is not cancellable.
+ * @param cancelInFlight If true, any existing effect with the same ID will be cancelled.
+ * @return ReduceResult containing the current state and the defined effect.
+ */
+fun <State, Action> State.withFlowEffect(
+    flow: Flow<Action>,
+    id: Any? = null,
+    cancelInFlight: Boolean = false,
+): ReduceResult<State, Action> =
+    ReduceResult(this, Effect.fromFlow(flow).maybeCancellable(id, cancelInFlight))
 
 /**
- * Used to indicate that a reducer does not have any effects other than mutating the state.
- * @return An empty list of effects.
+ * Returns a ReduceResult with the current state and a suspend function-based effect.
+ * The effect can be cancellable and is defined by the provided suspend function.
+ *
+ * @param id Optional ID for making the effect cancellable. If null, effect is not cancellable.
+ * @param cancelInFlight If true, any existing effect with the same ID will be cancelled.
+ * @param func The suspend function that defines the action to be executed as an effect.
+ * @return ReduceResult containing the current state and the defined effect.
  */
-fun noEffect(): List<Effect<Nothing>> =
-    emptyList()
+fun <State, Action : Any> State.withSuspendEffect(
+    id: Any? = null,
+    cancelInFlight: Boolean = false,
+    func: suspend () -> Action,
+): ReduceResult<State, Action> =
+    ReduceResult(this, Effect.fromSuspend(func).maybeCancellable(id, cancelInFlight))
+
+/**
+ * Returns a ReduceResult with the current state and a callback flow-based effect.
+ * The effect can be made cancellable and is defined by a callback block.
+ *
+ * @param id Optional ID to make the effect cancellable. If null, the effect is not cancellable.
+ * @param cancelInFlight If true, any existing effect with the same ID will be cancelled.
+ * @param block The callback block defining the flow of actions.
+ * @return ReduceResult containing the current state and the defined effect.
+ * @see kotlinx.coroutines.flow.callbackFlow
+ */
+@OptIn(ExperimentalTypeInference::class)
+fun <State, Action> State.withProducerEffect(
+    id: Any? = null,
+    cancelInFlight: Boolean = false,
```

Comment:
This combination of parameters allows for invalid state to be expressed `(null, true)`. This rustles my jimmies. We should probably combine them in a sum type instead:

```kotlin
sealed interface CancellableConfiguration {
    data class AllowMultiple(val id: Any) : CancellableConfiguration
    data class CancelInFlight(val id: Any) : CancellableConfiguration
}
```
Names not final. Then we can do:

```suggestion
    configuration: CancellableConfiguration? = null
```

Reply:
I like your idea of multiple overloads better!